src/librustc_mir/borrow_check/mod.rs

```diff
@@ -2373,23 +2371,71 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         &mut self,
         _: Context,
         (place, span): (&Place<'tcx>, Span),
+        borrow: &BorrowData<'tcx>,
+        scope_tree: &Rc<ScopeTree>,
         end_span: Option<Span>,
     ) {
+        let borrow_span = self.mir.source_info(borrow.location).span;
+
         let root_place = self.prefixes(place, PrefixSet::All).last().unwrap();
         let proper_span = match *root_place {
             Place::Local(local) => self.mir.local_decls[local].source_info.span,
             _ => span,
         };
-        let mut err = self.tcx
-            .path_does_not_live_long_enough(span, "borrowed value", Origin::Mir);
-        err.span_label(proper_span, "temporary value created here");
-        err.span_label(span, "temporary value dropped here while still borrowed");
-        err.note("consider using a `let` binding to increase its lifetime");
-
-        if let Some(end) = end_span {
-            err.span_label(end, "temporary value needs to live until here");
-        }
 
+        let name = &self.describe_place(place);
+        let (with_name, name) = match name {
+            Some(name) => (true, format!("`{}`", name)),
+            None => (false, "borrowed value".to_owned()),
+        };
+
+        let err_span = match borrow.region {
+            // See test ui/issue-46471-1.rs for this case.
+            RegionKind::ReScope(_) => span,
+            // See test ui/issue-46471.rs for this case.
+            RegionKind::ReEarlyBound(_) | RegionKind::ReFree(_) => proper_span,
+            // See test compile-fail/issue-36082.rs for this case.
+            _ if name == "borrowed value" => span,
+            // See test ui/issue-46472.rs for this case.
+            _ => borrow_span,
+        };
+
+        let mut err = self.tcx.path_does_not_live_long_enough(err_span, &name, Origin::Mir);
+        match borrow.region {
```

Comment:
There looks to be quite some connection between the span and the error message, so I would prefer to have 4 separate functions - 1 for each distinct error messagee